src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ResourceCodeDomSerializer.SerializationResourceManager.cs

```
@@ -154,20 +120,17 @@ private IResourceWriter Writer
         /// <summary>
         ///  The component serializer supports caching serialized outputs for speed.  It holds both a collection of statements as well as an opaque blob for resources.  This function adds data to that blob. The parameters to this function are the same as those to SetValue, or SetMetadata (when isMetadata is true).
         /// </summary>
-        private static void AddCacheEntry(IDesignerSerializationManager manager, string name, object value, bool isMetadata, bool forceInvariant, bool shouldSerializeValue, bool ensureInvariant)
+        private static void AddCacheEntry(IDesignerSerializationManager manager, string name, object? value, bool isMetadata, bool forceInvariant, bool shouldSerializeValue, bool ensureInvariant)
         {
-            if (manager.Context[typeof(ComponentCache.Entry)] is ComponentCache.Entry entry)
+            if (manager.TryGetContext(out ComponentCache.Entry? entry))
             {
-                ComponentCache.ResourceEntry re = new ComponentCache.ResourceEntry
-                {
-                    Name = name,
-                    Value = value,
-                    ForceInvariant = forceInvariant,
-                    ShouldSerializeValue = shouldSerializeValue,
-                    EnsureInvariant = ensureInvariant,
-                    PropertyDescriptor = (PropertyDescriptor)manager.Context[typeof(PropertyDescriptor)],
-                    ExpressionContext = (ExpressionContext)manager.Context[typeof(ExpressionContext)]
-                };
+                ComponentCache.ResourceEntry re = new ComponentCache.ResourceEntry(name,
+                    value,
+                    forceInvariant,
+                    shouldSerializeValue,
+                    ensureInvariant,
+                    manager.GetContext<PropertyDescriptor>()!,
+                    manager.GetContext<ExpressionContext>()!);
```

Please add comment of how do we know these are not null

---

Well, I don't know the actual reason... I just know the call sites of those properties always expect them not to be null as they are used as parameters in `manager.Context.Push` calls. That's another case of inconsistent checks I think

---

I see, it looks like these are used as arguments to `manager.Context.Push` only when `isMetadata = false`. So it seems that it is OK for `ResourceEntry.PropertyDescription` and `ResourceEntry.ExpressionContext` to be nullable and should be annotated as so. If these are null when `isMetadata = false`, we should actually throw `ArgumentException` here instead of later getting an `ArgumentNullException` from `manager.Context.Push` which might be confusing.

---

I've pushed something to address this. I've tried to stick to what you indicated. Let me know if this needs some more tweaking